internal/validator/handler.go

```diff
@@ -42,18 +44,45 @@ type handler struct {
 	producer *kafka.Producer
 	schema   *jsonschema.Schema
 	errors   chan error
+	msgChan  chan *kafka.Message
+	vChan    chan messageInfo
 }
 
 func (this *handler) onMessage(ctx context.Context, msg *kafka.Message) {
-	request := &messageModel.IngressValidationRequest{}
-	err := json.Unmarshal(msg.Value, request)
+	wg.Add(1)
 
-	if err != nil {
-		instrumentation.UnmarshallingError(ctx, err)
-		return
-	}
+	go func() {
+		defer wg.Done()
+
+		for {
+			if len(this.msgChan) >= cap(this.msgChan) {
+				continue
+			}
+			this.msgChan <- msg
+			break
+		}
+	}()
+	wg.Wait()
+}
+
+func (this *handler) fetchMessage(
+	ctx context.Context,
+) {
+	for {
+		request := &messageModel.IngressValidationRequest{}
+
+		select {
+		case msg := <-this.msgChan:
+			err := json.Unmarshal(msg.Value, request)
 
-	this.handleRequest(ctx, request)
+			if err != nil {
+				instrumentation.UnmarshallingError(ctx, err)
+				continue
+			}
+		}
+
+		go this.handleRequest(ctx, request)
```

Comment:
We should make it possible to limit the maximum concurrency of prefetching (i.e. how many "workers" can run in concurrently).

Reply:
Changed this to start a set number of goroutines instead of one for every message. Let me know if this is what you had in mind.